pkg/storageincentives/nodestatus.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package storageincentives
+
+import (
+	"context"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethersphere/bee/pkg/log"
+	"github.com/ethersphere/bee/pkg/settlement/swap/erc20"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/storageincentives/redistribution"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"math/big"
+	"sync"
+)
+
+const redistributionStatusKey = "redistribution_state"
+const loggerNameNode = "nodestatus"
+
+type NodeState struct {
+	stateStore     storage.StateStorer
+	erc20Service   erc20.Service
+	logger         log.Logger
+	overlay        swarm.Address
+	mtx            sync.Mutex
+	status         Status
+	currentBalance *big.Int
+	contract       redistribution.Contract
+}
+
+// Status provide internal status of the nodes in the redistribution game
+type Status struct {
+	Phase        PhaseType `json:"phase"`
+	State        State     `json:"state"`
+	Round        uint64    `json:"round"`
+	LastWonRound uint64    `json:"lastWonRound"`
+	Block        uint64    `json:"block"`
+	Reward       *big.Int  `json:"reward"`
+	Fees         *big.Int  `json:"fees"`
+}
+
+type State int
+
+const (
+	winner State = iota + 1
+	frozen
+	idle
+)
+
+func (s State) String() string {
+	switch s {
+	case winner:
+		return "winner"
+	case frozen:
+		return "frozen"
+	case idle:
+		return "idle"
+	default:
+		return "unknown"
+	}
+}
+
+func NewNode(logger log.Logger, stateStore storage.StateStorer, erc20Service erc20.Service, contract redistribution.Contract) NodeState {
+	return NodeState{
+		stateStore:     stateStore,
+		erc20Service:   erc20Service,
+		logger:         logger.WithName(loggerNameNode).Register(),
+		currentBalance: big.NewInt(0),
+		contract:       contract,
+		status: Status{
+			Round:  0,
+			Block:  0,
+			Reward: nil,
+			Fees:   nil,
+		},
+	}
+}
+
+func (n *NodeState) SetCurrentEvent(p PhaseType, r uint64, b uint64) {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+	n.status.Phase = p
+	n.status.Round = r
+	n.status.Block = b
+	n.status.State = idle
```

Comment:
if the node was frozen, this would remove the frozen state...

Reply:
Removed